src/visualizers-workbench/P5VisualizerTemplate.ts

```diff
@@ -0,0 +1,183 @@
+        if (this.params.stepSize.value <= 0) {
+            status.isValid = false
+            status.errors.push('Step size must be positive')
+        }
+
+        return status
+    }
+
+    setup() {
+        // the base setup function includes the `createCanvas()` call that must
+        // appear in every p5 setup function
+        //
+        //   https://p5js.org/reference/#/p5/createCanvas
+        //
+        super.setup()
+
+        const sketch = this.sketch
+
+        // initialize palette colors, chosen from the Numberscope site palette
+        this.bgColor = sketch.color(206, 212, 218)
+        this.textColor = sketch.color(128, 159, 255)
+        const outlineColor = sketch.color(51, 51, 255)
+
+        // set the stroke color and text alignment, which won't change from
+        // frame to frame. each setting method returns a reference to the
+        // sketch, so you can chain the methods as shown here
+        sketch.stroke(outlineColor).textAlign(sketch.CENTER, sketch.CENTER)
+
+        // start at the beginning of the sequence
+        this.index = this.seq.first
+    }
+
+    // this is where you draw your visualization! look at the examples and the
+    // p5 tutorials and reference to learn about what you can do
+    //
+    //   https://p5js.org/learn/
+    //   https://p5js.org/reference/
+    //
+    // you have to implement this method, even if it does nothing. your
+    // visualizer can't be loaded into Numberscope without it
+    draw() {
+        // accessing `this.sketch` triggers some consistency checks. by storing
+        // its value as a local constant, we avoid redundant checks
+        const sketch = this.sketch
+
+        // scale and center the coordinate system; scale the text to fit the
+        // canvas. calculations involving canvas dimensions need to be done
+        // every frame, because the dimensions can change at any time
+        const smallDim = Math.min(sketch.width, sketch.height)
+        sketch
+            .translate(0.5 * sketch.width, 0.5 * sketch.height)
+            .textSize(0.2 * smallDim)
+
+        // paint the background. for an animated visualizer, this is often done
+        // as the first drawing step in each frame, wiping out the previous
+        // frame and leaving a blank canvas to draw on
+        sketch.background(this.bgColor as p5.Color)
+
+        // print the current entry
+        const element = this.seq.getElement(this.index)
+        sketch.fill(this.textColor as p5.Color).text(element.toString(), 0, 0)
+
+        // draw a progress bar. drawing methods can be chained, just like
+        // setting methods
+        const progress = 1 - 7 / (7 + Math.sqrt(this.index - this.seq.first))
+        const barLength = 0.6 * smallDim
+        sketch
+            .fill(255, 128 + 128 * progress, 0)
+            .translate(-0.5 * barLength, 0.2 * smallDim)
+            .line(0, 0, barLength, 0)
+            .circle(progress * barLength, 0, 0.02 * smallDim)
+
```

Comment:
Ah, I had assumed you were using some slider widget that p5 provided, but now I see you are rolling your own. If there is a progress bar widget, I'd suggest using it. Presuming not, I'd make a long skinny horizontal rectangle outlined in your stroke color with the first `progress` fraction in the textcolor and the remainder in the bgcolor.

As for the progress fraction, why not use the usual linear fraction when this.seq.last is finite? I guess when it is Infinity, something like what you do is unavoidable (how'd you choose 7? It's a magic number, and should probably be explained and/or put in an aptly-named variable.)

Reply:
As far as I can tell, there is no built-in progress bar in p5.js. For everyone's amusement, here is a CSS and HTML-only very slick-looking progress bar, that does depend on a div outside the canvas that you would have to insert in the inhabit() function, making it too advanced (I think) for the template: https://editor.p5js.org/slow_izzm/sketches/ZXf_DolCG

Reply:
Oh, there is a slider widget! I've gone with an explicitly drawn progress bar, though, because I want to emphasize basic drawing and color-setting techniques. The code is more complicated than the line-and-circle progress bar, but I agree that the visual language is less confusing.

I've assigned a `barScale` constant for the magic number 7, which I choose so that the first few steps would look noticeably discrete, but wouldn't fill up a huge fraction of the progress bar.

> why not use the usual linear fraction when this.seq.last is finite?

This would certainly be better, but I'd like to keep the logic simple in the template.

Reply:
OK, we are converging here. I recommend
(A) as the drawing is now quite several operations and this is supposed to be a tutorial,  you add in some in-line comments like `\\ left (colored) portion of the progress bar` or what have you.
(B) in particular, I puzzled over the `128 + 128*progress`. Maybe add `\\ bar goes from red to yellow as we move along sequence`
(C) I get it about keeping the logic simple in this example, but right now it's at the expense of making the math/model of progress fairly obtuse. I strongly recommend creating a practical upper limit by say capping this.seq.last by this.seq.first + 200*stepSize (on the theory nobody is going to sit there and hit the right-arrow key more than 200 times) and then just use linear progress on the range from this.seq.first to the practical upper limit. That will be much more transparent than the current code to the unsophisticated reader of this example, and will not involve any more complex logic.

Reply:
I think there's a nice way to scale the progress bar without setting an arbitrary upper limit, and it would be a good way to demonstrate how to place and document algorithms that come with additional documentation. I vote to put this in a future PR.

Reply:
OK, I put in some (basically placeholder) commentary to this effect to model an additional documentation section, and referred to the point in a wrapup paragraph added to the making a visualizer overview. If you would like to beef this up and add some actual mathematical detail in a future PR, that'd be super. Resolving for now.